### ISSUE TITLE
FR:  Add option to utilize SSH certificates vs keys for Tailscale SSH.

### ISSUE THREAD

**Issue by Valien** [tailscale/tailscale#8732]
### What are you trying to do?

OpenSSH has had support for SSH based certificates since 2010 (see - https://www.openssh.com/txt/release-5.4). Setting up SSH certificates over key rotation has some nice advantages (auto-expiry, can't be tampered, etc) and reduces the potential of keys being leaked and re-used.

Other products leveraging SSH certifcates are Smallstep (https://smallstep.com/blog/use-ssh-certificates/) and Teleport (https://goteleport.com/blog/how-to-configure-ssh-certificate-based-authentication/)

If Tailscale offered the ability to use certs over keys (or even give users options) I think it would really make the core product stand out among other solutions out there that only leverage keys.

### How should we solve this?

Implement SSH-based certificate generation and revocation within Tailscale. Requires a CA to sign these certs. Something we could offer on users Tailnets possibly.

### What is the impact of not solving this?

No huge impact directly. The move in the industry is to leave behind static (or non-static) keys, secrets, and more. By going to a certificate-based system you are closing the gap towards passwordless and reducing the amount of keys out in the wild.

### Anything else?

Our page here - https://tailscale.com/tailscale-ssh/ - which is really nice, would really stand out if we were able to showcase cert-based access vs keys as we do it now.

**Comment by moljnir** [comment]
I'll second that; providing a ssh CA to enable the use of ssh certificates should be the default.

**Comment by DentonGentry** [comment]
The point of [Tailscale SSH](https://tailscale.com/tailscale-ssh/) is to authenticate using the tailnet's notion of a user's identity, not requiring a pre-shared key or shared secret.

Thought I understand there are systems out there which use certificates, could you explain why implementing certificate support would be preferable to use of the Tailscale identity?

**Comment by Valien** [comment]
We all know that managing PKI and SSH keys is not fun. Wrongly implemented can lead to attacks and access to resources when there should be no access.

Keys by their nature don't expire. They have to be manually rotated (or automatically rotated) but in reality very few orgs probably do this well. A user gets a key and uses it forever.

Even our Tailscale SSH docs touch on this:

> If a key is compromised, the device can [remove the keys on the device, re-install Tailscale](https://tailscale.com/kb/1069/uninstall/), and re-authenticate to generate and distribute new node and SSH host keys.

When using SSH-based certificates you can eliminate this step. Not really worried about a compromised key because if it's a certificate, then it will expire in a few minutes/hours, etc.  

I love how Tailscale SSH works. It's seemless and super easy to get going. I think that adding support for cert-based authentication could potentially appeal to orgs that are moving away from keys (even if it is handled by Tailscale) and going to a short-lived certificate-based system.

@awly might be able to comment on this more as I think he built some of the underlying stuff at our former company.

Personally, I'm a big fan of using certificates over keys. I auth once to my system. I get a cert and I can now SSH into any device that my RBAC rules allow me to based on the cert metadata. Then that cert expires at the end of the day and I don't have to worry about my local SSH key that is sitting on my laptop.  Now with Tailscale SSH we do it similarily in that there are no local keys (yay!) and handle the access via the users logged in identity (yay!) and ACL rules but maybe adding cert support will take the backend a step further than how it's handled now.

Just some additional thoughts on this! :)

**Comment by awly** [comment]
**TL;DR:** I think the current Tailscale SSH model is at par or better than certs, but it's not widely communicated or understood.

Tailscale SSH does not use client public key authentication in the classic sense.

In the classic setup, SSH server accepts an untrusted TCP connection, challenges client to authenticate (via password, public key, or certificate). Once the client authenticates using one of those methods, the session is established.

The incoming connection on the server is intercepted by `tailscaled` and authenticated at the WireGuard level. The SSH server skips user authentication altogether (using auth method `none`) because the underlying TCP connection is already trusted.

There are also SSH host keys in both setups that work roughly the same. AFAIK Tailscale only has those because it's a mandatory part of SSH handshake, they don't carry any security properties in our case. 

Here's a list of benefits of SSH certificates and their Tailscale couterparts:
* key distribution
  * certs rely on distributing one CA certificate and validation of all client certs using that
  * tailscale relies on fast distribution of tailnet node keys (note: these are not SSH client keys) via coordination server
* key revocation
  * certs rely on expiry to eventually invalidate old keys
  * tailscale relies on the coordination server immediately propagating node key removal on ACL update; also [check mode](https://tailscale.com/kb/1193/tailscale-ssh/#configure-tailscale-ssh-with-check-mode) can be used for implicit revocation based on time, just like with certs

I do agree that the compromised node rotation UX is not ideal (delete keys and reinstall tailscale). We can do better :)

What would be helpful:
1. Documentation on Tailscale SSH vs SSH certs
2. CLI command to regenerate node keys (`tailscale reset` or `tailscale up --rotate`?) for forced rotation

**Comment by bradfitz** [comment]
cc @WalterHub as potential documentation topic.
